File: userapp/flutter_application_1/lib/models/website_component.dart
import 'package:json_annotation/json_annotation.dart';
import 'component_registry.dart';

part 'website_component.g.dart';

@JsonSerializable()
class WebsiteComponent {
  final int id;
  final int websiteId;
  final ComponentRegistry componentRegistry;
  final String schemaData; // JSON string with customized values
  final int position;
  final bool visible;
  final DateTime createdAt;
  final DateTime updatedAt;

  WebsiteComponent({
    required this.id,
    required this.websiteId,
    required this.componentRegistry,
    required this.schemaData,
    required this.position,
    required this.visible,
    required this.createdAt,
    required this.updatedAt,
  });

  factory WebsiteComponent.fromJson(Map<String, dynamic> json) => _$WebsiteComponentFromJson(json);
  Map<String, dynamic> toJson() => _$WebsiteComponentToJson(this);
}

File: backend/src/main/java/com/ezweb/model/WebsiteComponent.java
package com.ezweb.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

@Entity
@Table(name = "website_components")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class WebsiteComponent {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "website_id", nullable = false)
    private Website website;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "component_registry_id", nullable = false)
    private ComponentRegistry componentRegistry;

    @Column(name = "schema_data", columnDefinition = "TEXT", nullable = false)
    private String schemaData; // JSON data with customized values

    @Column(nullable = false)
    private Integer position = 0; // Order of component on page

    @Column(nullable = false)
    private Boolean visible = true;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}
